### ISSUE TITLE
deleting a consumer hang forever 

### ISSUE THREAD

**Issue by horchi** [edenhill/librdkafka#3182]
Description
===========
I have a consumer created like this:
`if (!(consumer = RdKafka::KafkaConsumer::create(conf, errstr)))`

now I subscribe to one topic and receive some messages. After a view messages I like to stop and delete the consumer, therefore I unassign and delete the consumer:
```
 consumer->unassign();
 consumer->stop();
 delete consumer;
```
Now it stochastic (not each time) hang forever in the delete.
The only way to get the control back is to kill the entire process.
Here is the back trace of the delete
```
(gdb) bt
#0  0x00007fae0846af47 in pthread_join (threadid=threadid@entry=140385340847872, thread_return=thread_return@entry=0x7ffe4c99f678) at pthread_join.c:90
#1  0x00007fae0589de72 in thrd_join (thr=thr@entry=140385340847872, res=res@entry=0x7ffe4c99f69c) at tinycthread.c:692
#2  0x00007fae0581d3b4 in rd_kafka_destroy_app (rk=0x244acc0, flags=<optimized out>, flags@entry=8) at rdkafka.c:1112
#3  0x00007fae0581d735 in rd_kafka_destroy_flags (rk=<optimized out>, flags=flags@entry=8) at rdkafka.c:1129
#4  0x00007fae07342552 in RdKafka::KafkaConsumerImpl::~KafkaConsumerImpl (this=0x24273c0, __in_chrg=<optimized out>, __vtt_parm=<optimized out>) at rdkafkacpp_int.h:1113
#5  RdKafka::KafkaConsumerImpl::~KafkaConsumerImpl (this=0x24273c0, __in_chrg=<optimized out>, __vtt_parm=<optimized out>) at rdkafkacpp_int.h:1114
#6  0x00000000004d6ac3 in KafkaClient::disconnect (this=0x2426020) at kafka.cc:91
```

Checklist
=========
 - [x] librdkafka version: 1.6.0 - commit-id 5aea437
 - [x] Apache Kafka version: `Kafka version: 2.6.0
 - [x] librdkafka client configuration:  socket.timeout.ms=5000;  fetch.wait.max.ms=5; fetch.min.bytes=1024; enable.auto.commit=false
 - [x] Operating system: CentOS Linux release 7.6.1810 (Core)
 - [] Provide logs (with `debug=..` as necessary) from librdkafka
 - [] Provide broker log excerpts
 - [x] Critical issue

Thanks!

**Comment by horchi** [comment]
the thread is hang in this loop in rdkafka_broker.c:5089:

```
	while (!rd_kafka_broker_terminating(rkb)) {
                int backoff;
                int r;

        redo:
                switch (rkb->rkb_state)
                {
                case RD_KAFKA_BROKER_STATE_INIT:
                        /* Check if there is demand for a connection
                         * to this broker, if so jump to TRY_CONNECT state. */
                        if (!rd_kafka_broker_needs_connection(rkb)) {
                                rd_kafka_broker_serve(rkb,
                                                      rd_kafka_max_block_ms);
                                break;
                        }
```

**Comment by Phantomape** [comment]
I got the same issue here, any idea?

**Comment by horchi** [comment]
yes in between I have an idea, the lib seems a bit touchy if an object isn't proper deleted before calling consumers destruction. 
I debugged into the lib and it sees to be a monitoring of some objects and blocking the exit of the while loop above.

I my case it may be a un-deleted message object got from RdKafka::KafkaConsumer::consume. In some cases the last received messages was not processes (therefore even not commited). After I add the message deletion the problem not appeared anymore, but I have to wait until it appear only once in a few hours.
Here I found some documentation about it:
https://github.com/edenhill/librdkafka/blob/master/INTRODUCTION.md#termination
It would be nice to get some log information instead of infinite blocking the process? Even a delete of the message and a possible subsequent cash may easier to debug?
